relay-server/src/services/project.rs

```diff
@@ -1053,9 +1054,37 @@ impl Project {
             Ok(current_limits.check_with_quotas(quotas, item_scoping))
         });
 
+        let check_nested_spans = state
+            .as_ref()
+            .is_some_and(|s| s.has_feature(Feature::ExtractSpansFromEvent));
+
+        #[derive(Debug, Deserialize)]
+        struct PartialEvent {
+            spans: SeqCount,
+        }
+
+        // We compute the number of nested spans before rate limiting since we will remove the
+        // transaction item in case it is rate limited.
+        let spans = if check_nested_spans {
+            envelope
+                .envelope()
+                .items()
+                .find(|i| *i.ty() == ItemType::Transaction && !i.spans_extracted())
+                .and_then(|i| serde_json::from_slice::<PartialEvent>(&i.payload()).ok())
+                .map_or(0, |p| p.spans.0)
+        } else {
+            0
```

Comment:
Not very nice, but I wanted to avoid unnecessary work in case the flag is not set.